From 0cea46c309d6af275c38a335a2c3c05b43f4d8b1 Mon Sep 17 00:00:00 2001
From: Philippe Gerum <philippe.gerum@exail.com>
Date: Tue, 22 Apr 2025 15:38:07 +0200
Subject: [PATCH 0161/1013] net: dovetail: fix skb-related misnomers

This patch fixes two serious misnomers which are error-prone to any
newcomer to the oob netstack:

- an [oob_]skb is a buffer shell, not a buffer head which refers to
  the linear storage space conveying the start of the payload instead.

- skb_has_oob_storage() conveys the wrong message because this flag
  should be raised for any buffer which storage area is managed by the
  companion core, which is distinct from saying that such storage has
  to belong to an oob pool. Typically, we may have storage obtained
  from an oob-capable page pool used by in-band code, and which
  lifetime is solely controlled in such context, independently of the
  companion core running an oob netstack.

Signed-off-by: Philippe Gerum <philippe.gerum@exail.com>
---
 include/linux/skbuff.h | 43 +++++++++++++++++++++---------------------
 net/core/skbuff.c      |  6 +++---
 2 files changed, 25 insertions(+), 24 deletions(-)

diff --git a/include/linux/skbuff.h b/include/linux/skbuff.h
index c82c295dd1d0..144fface3e7e 100644
--- a/include/linux/skbuff.h
+++ b/include/linux/skbuff.h
@@ -520,8 +520,9 @@ enum {
 	 */
 	SKBFL_MANAGED_FRAG_REFS = BIT(4),
 
-	/* data lives into an oob-enabled storage. */
-	SKBFL_OOB_STORAGE = BIT(5),
+	/* data storage is managed by a companion core operating from
+	 * the oob stage. */
+	SKBFL_OOB_MANAGED = BIT(5),
 };
 
 #define SKBFL_ZEROCOPY_FRAG	(SKBFL_ZEROCOPY_ENABLE | SKBFL_SHARED_FRAG)
@@ -943,9 +944,6 @@ struct sk_buff {
 				head_frag:1,
 				pfmemalloc:1,
 				pp_recycle:1; /* page_pool recycle indicator */
-#ifdef CONFIG_NET_OOB
-	__u8			oob_head:1;
-#endif
 #ifdef CONFIG_SKB_EXTENSIONS
 	__u8			active_extensions;
 #endif
@@ -992,6 +990,9 @@ struct sk_buff {
 	__u8			ndisc_nodetype:2;
 #endif
 
+#ifdef CONFIG_NET_OOB
+	__u8			oob:1;
+#endif
 #if IS_ENABLED(CONFIG_IP_VS)
 	__u8			ipvs_property:1;
 #endif
@@ -5264,37 +5265,37 @@ struct sk_buff *get_oob_skb(void);
 void put_oob_skb(struct sk_buff *skb);
 
 /**
- * skb_is_oob - Whether the buffer was allocated from the oob
- * pool. Caution: this is distinct from the skb conveying oob storage,
- * see skb_has_oob_storage().
+ * skb_is_oob - Whether the buffer shell was allocated from the oob
+ * pool. Caution: this is distinct from a skb which storage is managed
+ * by a companion core, see skb_is_oob_managed().
  */
 static inline bool skb_is_oob(const struct sk_buff *skb)
 {
-	return skb->oob_head;
+	return skb->oob;
 }
 
 static inline void skb_mark_oob(struct sk_buff *skb)
 {
-	skb->oob_head = 1;
+	skb->oob = 1;
 }
 
 /**
- * skb_has_oob_storage - Whether the skb data lives in an oob-enabled
- * page pool.
+ * skb_is_oob_managed - Whether the skb data is managed by a companion
+ * core.
  */
-static inline bool skb_has_oob_storage(const struct sk_buff *skb)
+static inline bool skb_is_oob_managed(const struct sk_buff *skb)
 {
-	return skb_shinfo(skb)->flags & SKBFL_OOB_STORAGE;
+	return skb_shinfo(skb)->flags & SKBFL_OOB_MANAGED;
 }
 
-static inline void skb_mark_oob_storage(struct sk_buff *skb)
+static inline void skb_mark_oob_managed(struct sk_buff *skb)
 {
-	skb_shinfo(skb)->flags |= SKBFL_OOB_STORAGE;
+	skb_shinfo(skb)->flags |= SKBFL_OOB_MANAGED;
 }
 
-static inline dma_addr_t skb_oob_storage_addr(const struct sk_buff *skb)
+static inline dma_addr_t skb_oob_dma_addr(const struct sk_buff *skb)
 {
-	if (!skb->head || !skb_has_oob_storage(skb) || !skb->pp_recycle)
+	if (!skb->head || !skb_is_oob_managed(skb) || !skb->pp_recycle)
 		return DMA_MAPPING_ERROR;
 
 	return page_pool_get_dma_addr(virt_to_page(skb->head));
@@ -5316,7 +5317,7 @@ static inline void finalize_skb_inband(struct sk_buff *skb)
  */
 static inline void __skb_inband_clone(struct sk_buff *skb)
 {
-	skb->oob_head = 0;
+	skb->oob = 0;
 }
 
 /**
@@ -5353,12 +5354,12 @@ static inline bool __skb_oob_free_head(struct sk_buff *skb)
 	return false;
 }
 
-static inline bool skb_has_oob_storage(const struct sk_buff *skb)
+static inline bool skb_is_oob_managed(const struct sk_buff *skb)
 {
 	return false;
 }
 
-static inline dma_addr_t skb_oob_storage_addr(const struct sk_buff *skb)
+static inline dma_addr_t skb_oob_dma_addr(const struct sk_buff *skb)
 {
 	return DMA_MAPPING_ERROR;
 }
diff --git a/net/core/skbuff.c b/net/core/skbuff.c
index 43d1fd518a4d..bb6cd03c1ded 100644
--- a/net/core/skbuff.c
+++ b/net/core/skbuff.c
@@ -601,10 +601,10 @@ bool recycle_skb_oob(struct sk_buff *skb)
 {
 	/*
 	 * Hand the buffer release over the out-of-band core either if
-	 * the latter owns the skb data or we are currently running
-	 * oob.
+	 * the latter manages the data storage or we are currently
+	 * running oob.
 	 */
-	if (running_oob() || skb_has_oob_storage(skb)) {
+	if (running_oob() || skb_is_oob_managed(skb)) {
 		free_skb_oob(skb);
 		return true;
 	}
-- 
2.30.2

